src/coreclr/src/pal/inc/pal_mstypes.h

```diff
@@ -214,14 +214,16 @@ extern "C" {
 #ifndef PAL_IMPLEMENTATION
 
 // OS X already defines these types in 64 bit
-#if !defined(TARGET_OSX)
+#ifndef TARGET_OSX
 typedef __int64 int64_t;
 typedef unsigned __int64 uint64_t;
 typedef __int32 int32_t;
 typedef unsigned __int32 uint32_t;
 typedef __int16 int16_t;
 typedef unsigned __int16 uint16_t;
+#if !defined(TARGET_ANDROID) || defined(FEATURE_NO_HOST)
```

Comment:
Using this define inside PAL headers does not sounds right. How is this related to `FEATURE_NO_HOST`?

Reply:
Without this whole condition, the error was:

```
  [ 69%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/objmgr/shmobjectmanager.cpp.o
  [ 69%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/safecrt/makepath_s.cpp.o
  In file included from /datadrive/projects/runtime4/src/coreclr/src/pal/src/safecrt/makepath_s.cpp:20:
  In file included from /datadrive/projects/runtime4/src/coreclr/src/pal/src/safecrt/mbusafecrt_internal.h:21:
  /datadrive/projects/runtime4/src/coreclr/src/pal/inc/pal_mstypes.h:226:16: error: typedef redefinition with different types ('char' vs '__int8_t' (aka 'signed char'))
  typedef __int8 int8_t;
                 ^
  /datadrive/projects/android-arm64-rootfs/android-rootfs/android-ndk-r21/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/stdint.h:58:23: note: previous definition is here
  typedef __int8_t      int8_t;
                        ^
  1 error generated.
  src/pal/src/CMakeFiles/coreclrpal.dir/build.make:790: recipe for target 'src/pal/src/CMakeFiles/coreclrpal.dir/safecrt/makepath_s.cpp.o' failed
  make[2]: *** [src/pal/src/CMakeFiles/coreclrpal.dir/safecrt/makepath_s.cpp.o] Error 1
  CMakeFiles/Makefile2:252: recipe for target 'src/pal/src/CMakeFiles/coreclrpal.dir/all' failed
  make[1]: *** [src/pal/src/CMakeFiles/coreclrpal.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
```

and if with the `|| defined(FEATURE_NO_HOST)` part, the error was:

```
  [ 85%] Building CXX object src/vm/crossgen/CMakeFiles/cee_crossgen.dir/__/stubcache.cpp.o
  [ 85%] Building CXX object src/vm/crossgen/CMakeFiles/cee_crossgen.dir/__/stubgen.cpp.o
  [ 85%] Building CXX object src/vm/crossgen/CMakeFiles/cee_crossgen.dir/__/stublink.cpp.o
  Scanning dependencies of target clrjit
  [ 85%] Building CXX object src/jit/standalone/CMakeFiles/clrjit.dir/__/alloc.cpp.o
  In file included from /datadrive/projects/runtime4/src/coreclr/src/jit/alloc.cpp:5:
  In file included from /datadrive/projects/runtime4/src/coreclr/src/jit/jitpch.h:19:
  In file included from /datadrive/projects/runtime4/src/coreclr/src/jit/jit.h:806:
  In file included from /datadrive/projects/runtime4/src/coreclr/src/jit/compiler.h:30:
  /datadrive/projects/runtime4/src/coreclr/src/jit/gentree.h:2830:17: error: unknown type name 'int8_t'; did you mean 'uint8_t'?
          return (int8_t)val == val;
                  ^~~~~~
                  uint8_t
  /datadrive/projects/runtime4/src/coreclr/src/pal/inc/pal_mstypes.h:228:25: note: 'uint8_t' declared here
  typedef unsigned __int8 uint8_t;
                          ^
  [ 85%] Building CXX object src/vm/crossgen/CMakeFiles/cee_crossgen.dir/__/typectxt.cpp.o
  1 error generated.
  make[2]: *** [src/jit/standalone/CMakeFiles/clrjit.dir/__/alloc.cpp.o] Error 1
  src/jit/standalone/CMakeFiles/clrjit.dir/build.make:62: recipe for target 'src/jit/standalone/CMakeFiles/clrjit.dir/__/alloc.cpp.o' failed
  CMakeFiles/Makefile2:1301: recipe for target 'src/jit/standalone/CMakeFiles/clrjit.dir/all' failed
  make[1]: *** [src/jit/standalone/CMakeFiles/clrjit.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
```

Reply:
@jkotas, there are also some preexisting similar #if[def] conditions in this header. If you could suggest a better approach, I can try that.

Reply:
We have two defines for this:

- `PAL_STDCPP_COMPAT` This define means: make PAL headers compatible with standard system headers.
- `PAL_IMPLEMENTATION` This means: PAL headers are included for PAL implementation; avoid stuff that conflicts with system headers required for PAL implementation.

I think that the fix for this issue should be some combination of these two in the right places.

Reply:
Thank you for details. Is it correct to put it this way that from the first log trace above, this conflict should not happen:

```
  /datadrive/projects/runtime4/src/coreclr/src/pal/inc/pal_mstypes.h:226:16: error: typedef redefinition with different types ('char' vs '__int8_t' (aka 'signed char'))
  typedef __int8 int8_t;
                 ^
  /datadrive/projects/android-arm64-rootfs/android-rootfs/android-ndk-r21/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/stdint.h:58:23: note: previous definition is here
  typedef __int8_t      int8_t;
```

i.e. the compilation at this point should not include `stdint.h` header? At this point, we are under `#ifndef` for both `PAL_STDCPP_COMPAT` and `PAL_IMPLEMENTATION`. NDK's `stdint.h` has `int8_t` defined as `unsigned char`, while we are trying to redefine it as `char`. I am trying to determine whether to investigate the compiler flags, or work on the control flow.

Reply:
@am11  It seems that the problem is that the safecrt compilation doesn't have the PAL_IMPLEMENTATION defined. The safecrt is kind of a strange part of PAL as all the other parts of PAL include palinternal.h that defines PAL_IMPLEMENTATION.
It seems that you could fix it by definining PAL_IMPLEMENTATION in mbusafecrt_internal.h

Reply:
Thank you @janvorli. This is the cleanest fix. I have tested and it works. After making other requested changes, I will push.